scripts/sanitycheck

```diff
@@ -1948,166 +1560,606 @@ class TestInstance:
+                self.suite.total_done, total_tests_width, self.suite.total_tests, instance.platform.name,
+                instance.testcase.name, status, handler_type))
+        else:
+                sys.stdout.write(
+                    "\rtotal complete: %s%4d/%4d%s  %2d%%  failed: %s%4d%s" %
+                    (COLOR_GREEN, self.suite.total_done, self.suite.total_tests, COLOR_NORMAL,
+                     int((float(self.suite.total_done) / self.suite.total_tests) * 100),
+                     COLOR_RED if self.suite.total_failed > 0 else COLOR_NORMAL, self.suite.total_failed,
+                     COLOR_NORMAL))
+        sys.stdout.flush()
+
+    def cmake(self):
+
+        instance = self.instance
+        args = self.testcase.extra_args[:]
+
+        if instance.handler:
+            args += instance.handler.args
+
+        # merge overlay files into one variable
+        overlays = ""
+        idx = 0
+        for arg in args:
+            match = re.search('OVERLAY_CONFIG="(.*)"', arg)
+            if match:
+                overlays += match.group(1)
+                del args[idx]
+                idx += 1
+
+        if len(self.testcase.extra_configs) > 0 or options.coverage:
+            args.append("OVERLAY_CONFIG=\"%s %s\"" %(overlays,
+                        os.path.join(instance.build_dir,
+                                     "sanitycheck", "testcase_extra.conf")))
+
+        results = self.run_cmake(args)
+
+        return results
+
+    def build(self):
+        results = self.run_build(['--build', self.build_dir])
+        return results
+
+    def run(self):
+        instance = self.instance
+
+        instance.handler.handle()
+        if instance.handler.type_str == "qemu":
+            verbose("Running %s (%s)" %(instance.name, instance.handler.type_str))
+
+            start_time = time.time()
+            command = [get_generator()[0]]
+            command += ["-C", self.build_dir, "run"]
+
+            with subprocess.Popen(command, stdout=subprocess.PIPE, stderr=subprocess.PIPE, cwd=self.build_dir) as proc:
+                verbose("Spawning QEMUHandler Thread for %s" % instance.name)
+                proc.wait()
+                self.returncode = proc.returncode
+
+            instance.metrics["handler_time"] = time.time() - start_time
+
+
+pipeline = queue.LifoQueue()
+
+class BoundedExecutor(concurrent.futures.ThreadPoolExecutor):
+    """BoundedExecutor behaves as a ThreadPoolExecutor which will block on
+    calls to submit() once the limit given as "bound" work items are queued for
+    execution.
+    :param bound: Integer - the maximum number of items in the work queue
+    :param max_workers: Integer - the size of the thread pool
+    """
+    def __init__(self, bound, max_workers, **kwargs):
+        super().__init__(max_workers, kwargs)
+        #self.executor = ThreadPoolExecutor(max_workers=max_workers)
+        self.semaphore = BoundedSemaphore(bound + max_workers)
```

Comment:
There's no need for this semaphore, nor to derive this class, nor for a "feeder" thread. This all  comes for free like this:
```python
stageN_workers = futures.ProcessPoolExecutor(max_workers=50)
stageN_completed = queue.Queue()
for some_job...
    # Submitting all of them is instantaneous.
    # They start only when there's a free worker 
    future = stageN_workers.submit(some_job,...) 
    future.add_done_callback(lambda f : stageN_completed.put(f))

# stage N+1
for .. in ...
    one_job_results = stageN_completed.get(timeout=1000)
```

Fully functional demo at https://gist.github.com/marc-hb/b537d30fd3d4fc0d506eff1b66f745e6

Reply:
make this work with sanitycheck and submit a patch :)
I spent a lot of time on this to get it to work the way it is now, so optimizations are secondary priority right now..

Reply:
Fair enough!
I doubt this would optimise anything, it's a simplification to spend less... developer time :-) not machine time.